# **ASSIGNMENT 2 - DIFFERENT ENSEMBLE LEARNING TECHNIQUES**

In [2]:
import numpy as np
import pandas as pd

In [3]:
# READING THE DATA AND FINDING NaN VALUES
data=pd.read_csv('/content/train_loan.csv')
data.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [4]:
data.isna().sum()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [5]:
# dropping the variable 'Loan_ID' since it has no significance
data.drop('Loan_ID',axis=1,inplace=True)

In [6]:
data.head()

Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
0   Male      No          0      Graduate            No             5849   
1   Male     Yes          1      Graduate            No             4583   
2   Male     Yes          0      Graduate           Yes             3000   
3   Male     Yes          0  Not Graduate            No             2583   
4   Male      No          0      Graduate            No             6000   

   CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                0.0         NaN             360.0             1.0   
1             1508.0       128.0             360.0             1.0   
2                0.0        66.0             360.0             1.0   
3             2358.0       120.0             360.0             1.0   
4                0.0       141.0             360.0             1.0   

  Property_Area Loan_Status  
0         Urban           Y  
1         Rural           N  
2         Urban           Y  
3         Urban           Y  
4         Urban           Y

In [7]:
#finding the categorical values
data.dtypes

Gender                object
Married               object
Dependents            object
Education             object
Self_Employed         object
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area         object
Loan_Status           object
dtype: object

In [8]:
# since Dependents contain values with + it is replaced
data['Dependents']=data['Dependents'].str.replace('+','')
data['Dependents'] = pd.to_numeric(data['Dependents'], errors='coerce').fillna(0).astype(int)

<ipython-input-8-28e91b24083b>:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data['Dependents']=data['Dependents'].str.replace('+','')


In [9]:
data.columns

Index(['Gender', 'Married', 'Dependents', 'Education', 'Self_Employed',
       'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area', 'Loan_Status'],
      dtype='object')

In [10]:
# Handling the missing values in the numerical variables-Dependents,LoanAmount,Loan_Amount_Term,Credit_History
data['Dependents'].fillna(data['Dependents'].median(), inplace = True)
data['LoanAmount'].fillna(data['LoanAmount'].median(), inplace = True)
data['Loan_Amount_Term'].fillna(data['Loan_Amount_Term'].median(), inplace = True)
data['Credit_History'].fillna(data['Credit_History'].median(), inplace = True)

In [11]:
data.describe()

Dependents  ApplicantIncome  CoapplicantIncome  LoanAmount  \
count  614.000000       614.000000         614.000000  614.000000   
mean     0.744300      5403.459283        1621.245798  145.752443   
std      1.009623      6109.041673        2926.248369   84.107233   
min      0.000000       150.000000           0.000000    9.000000   
25%      0.000000      2877.500000           0.000000  100.250000   
50%      0.000000      3812.500000        1188.500000  128.000000   
75%      1.000000      5795.000000        2297.250000  164.750000   
max      3.000000     81000.000000       41667.000000  700.000000   

       Loan_Amount_Term  Credit_History  
count        614.000000      614.000000  
mean         342.410423        0.855049  
std           64.428629        0.352339  
min           12.000000        0.000000  
25%          360.000000        1.000000  
50%          360.000000        1.000000  
75%          360.000000        1.000000  
max          480.000000        1.000000

In [12]:
data.isna().sum()

Gender               13
Married               3
Dependents            0
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount            0
Loan_Amount_Term      0
Credit_History        0
Property_Area         0
Loan_Status           0
dtype: int64

In [13]:
# Encoding
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
data.Gender=le.fit_transform(data.Gender)
data.Married=le.fit_transform(data.Married)
data.Education=le.fit_transform(data.Education)
data.Self_Employed=le.fit_transform(data.Self_Employed)
data.Property_Area=le.fit_transform(data.Property_Area)
data.Loan_Status=le.fit_transform(data.Loan_Status)
data.head()

Gender  Married  Dependents  Education  Self_Employed  ApplicantIncome  \
0       1        0           0          0              0             5849   
1       1        1           1          0              0             4583   
2       1        1           0          0              1             3000   
3       1        1           0          1              0             2583   
4       1        0           0          0              0             6000   

   CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                0.0       128.0             360.0             1.0   
1             1508.0       128.0             360.0             1.0   
2                0.0        66.0             360.0             1.0   
3             2358.0       120.0             360.0             1.0   
4                0.0       141.0             360.0             1.0   

   Property_Area  Loan_Status  
0              2            1  
1              0            0  
2              2            1  
3              2            1  
4              2            1

In [14]:
data.dtypes

Gender                 int64
Married                int64
Dependents             int64
Education              int64
Self_Employed          int64
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area          int64
Loan_Status            int64
dtype: object

In [15]:
data.isna().sum()

Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
Loan_Status          0
dtype: int64

All the categorical values are encoded and NaN are filled.
Now splitting the target column **"Loan_Status"**

In [16]:
x=data.drop('Loan_Status',axis=1)
y=data['Loan_Status']

In [17]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=42)

**RANDOM FOREST**

In [18]:
from sklearn.ensemble import RandomForestClassifier
rf_clf=RandomForestClassifier(n_estimators=150,random_state=42)
rf_clf.fit(x_train,y_train)
y_pred_rf=rf_clf.predict(x_test)
from sklearn import metrics
print('Accuracy = ',metrics.accuracy_score(y_test,y_pred_rf))

Accuracy =  0.7723577235772358


**BAGGING CLASSIFIER USING DECISION TREE**

In [19]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier
dct_clf=DecisionTreeClassifier()
bag_clf=BaggingClassifier(base_estimator=dct_clf,n_estimators=100,random_state=42)
bag_clf.fit(x_train,y_train)
y_pred_bag=bag_clf.predict(x_test)
print('Accuracy = ',metrics.accuracy_score(y_test,y_pred_bag))

/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


Accuracy =  0.7642276422764228


**ADABOOST CLASSIFIER**

In [20]:
from sklearn.ensemble import AdaBoostClassifier
ada_clf=AdaBoostClassifier(n_estimators=150,random_state=42)
ada_clf.fit(x_train,y_train)
y_pred_ada=ada_clf.predict(x_test)
print('Accuracy = ',metrics.accuracy_score(y_test,y_pred_ada))

Accuracy =  0.7479674796747967


**GRADIENT BOOSTING**

In [21]:
from sklearn.ensemble import GradientBoostingClassifier
grad_clf=GradientBoostingClassifier(n_estimators=40,max_depth=3,random_state=42)
grad_clf.fit(x_train,y_train)
y_pred_grad=grad_clf.predict(x_test)
print('Accuracy = ',metrics.accuracy_score(y_test,y_pred_grad))

Accuracy =  0.7886178861788617


**XGBOOSTING**

In [22]:
import xgboost as xgb
xgb_clf=xgb.XGBClassifier(n_estimators=40,max_depth=3,random_state=42)
xgb_clf.fit(x_train,y_train)
y_pred_xgb=xgb_clf.predict(x_test)
print('Accuracy = ',metrics.accuracy_score(y_test,y_pred_xgb))

Accuracy =  0.7642276422764228


**FINAL OUTCOME**
Accuracy of : RandomForestClassifier(77.2%), GradientBoosting(78.9%), AdaBoost(74.8%), BaggingClassifier(76.4%), XGBoosting(76.4%).
**GradientBoosting** has better accuracy.